bluesky/callbacks/tiled_writer.py

```diff
@@ -0,0 +1,165 @@
+import numpy as np
+import pandas as pd
+from event_model import DocumentRouter, RunRouter
+from tiled.client import from_profile, from_uri
+from tiled.structures.array import ArrayStructure, BuiltinDtype
+from tiled.structures.core import Spec, StructureFamily
+from tiled.structures.data_source import Asset, DataSource, Management
+from tiled.structures.table import TableStructure
+
+MIMETYPE_LOOKUP = {
+    "hdf5": "application/x-hdf5",
+    "ADHDF5_SWMR_STREAM": "application/x-hdf5",
+    "AD_HDF5_SWMR_SLICE": "application/x-hdf5",
+}
+
+
+class TiledWriter:
+    "Write metadata and data from Bluesky documents into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self._run_router = RunRouter([self._factory])
+
+    def _factory(self, name, doc):
+        return [_RunWriter(self.client)], []
+
+    @classmethod
+    def from_uri(cls, uri, **kwargs):
+        client = from_uri(uri, **kwargs)
+        return cls(client)
+
+    @classmethod
+    def from_profile(cls, profile, **kwargs):
+        client = from_profile(profile, **kwargs)
+        return cls(client)
+
+    def __call__(self, name, doc):
+        self._run_router(name, doc)
+
+
+class _RunWriter(DocumentRouter):
+    "Write the document from one Bluesky Run into Tiled."
+
+    def __init__(self, client):
+        self.client = client
+        self.node = None
+        self._descriptor_nodes = {}  # references to descriptor containers by uid's
+        self._SR_nodes = {}
+        self._SR_cache = {}
+
+    def start(self, doc):
+        self.node = self.client.create_container(
+            key=doc["uid"],
+            metadata={"start": doc},
+            specs=[Spec("BlueskyRun", version="1.0")],
+        )
+
+    def stop(self, doc):
+        metadata = dict(self.node.metadata) | {"stop": doc}
+        self.node.update_metadata(metadata=metadata)
+
+    def descriptor(self, doc):
+        descriptor_node = self.node.create_container(key=doc["name"], metadata=doc)
```

Comment:
In the original, group-coded draft of this, we intentionally made a simplifying assumption here: one Event Descriptor document per stream: i.e. we'll only see one such document with a given `name` key.

In general, the Event Descriptor with a given `name` can be re-emitted if a configuration setting (e.g. exposure time) changes. However, the `data_keys` are _not_ allowed to change. (The bundler code ensures this.)

To start, I might suggest following the structure we use in `databroker.mongo_normalized`, which is to place the Event Descriptor(s) in a list, under the key `descriptors`, as in `metadata={"descriptors": [doc]}` and then appending any additional descriptors with the same `name` to this list. That would be sufficient to have something that works.

In the future, pulling apart the changeable parts and the fixed parts (`data_keys`) may be useful. Needs more thought and dicussion.

Reply:
It could be the same file or a different file. It depends on the details of how the detector I/O works, whether it would open a new file or not. So, as far as Bluesky and Tiled go, we have to handle both possibilities.

If there _is_ a new file, there will be a new Stream Resource (and, in it, a new filepath). Given that, do we need to store this descriptor reference? I'm not sure it's needed, but I'm not sure it's _not_ needed....

> Would it make sense to have this list of descriptors stored in the top-most container (along with the Start/Stop messages) instead of its own node?

I think not, but let's examine this because I'm actually not sure what the best arguments are, and maybe there are better arguments in the other directions. Off the top of my head:

- It's easier to mix things than un-mix them, so I have a general impulse to place metadata on the specific thing it describes, in this case "the stream". Consumers can always hoist and flatten.
- The Event Descriptors can be large. For someone who is looping through a large number of scans and displaying basic metadata, it helps performance to only download the Start and Stop documents that describe the whole thing and defer downloading information about the streams.

Reply:
Thank you for explaining this, Dan. I had another thought about this last night, and I agree, mixing descriptors and start/stop messages is probably a bad idea.

I'm trying to get a better understanding of the underlying data structure and possible use cases, and so far I have the following image in my head.
![Bluesky_Messages_Relations](https://github.com/bluesky/bluesky/assets/15040858/f6fc5349-6b49-4806-9a20-cb6e4aff8896)

It seems to me that if different parts of the same stream resource (different Stream Datums: either rows of an HDF5 file, sets of tiffs from a folder, etc) can be related to different descriptors, we need some way to keep track of their indices _within_ the descriptor container, so paths like `/run/descriptor/dataset?slice=5:10` would resolve to the correct rows in the file, which could be e.g. `205:210`, or maybe not even be consecutive. I think `seq_nums` and `indices` in StreamDatum serve this purpose, but currently I am not storing them anywhere. (Should there be a lookup table or something like that?)
On the other hand, can we rely on individual Stream Datums alone to determine this correspondence between resource and descriptor indices? Do we need any kind of validation?